pyro/optim/optim.py

```diff
@@ -40,14 +43,19 @@ def __call__(self, params,  *args, **kwargs):
         Do an optimization step for each param in params. If a given param has never been seen before,
         initialize an optimizer for it.
         """
-
+        if hasattr(self, 'pt_scheduler_constructor'):
+            kwargs['epoch'] = self.epoch
         for p in params:
             # if we have not seen this param before, we instantiate and optim object to deal with it
             if p not in self.optim_objs:
                 # get our constructor arguments
                 def_optim_dict = self._get_optim_args(p)
                 # create a single optim object for that param
-                self.optim_objs[p] = self.pt_optim_constructor([p], **def_optim_dict)
+                optim = self.pt_optim_constructor([p], **def_optim_dict)
+                # PyroOptim is a scheduler
+                if hasattr(self, 'pt_scheduler_constructor'):
```

Comment:
nit: `hasattr` smells bad; does it make sense to define two methods
```py
class PyroOptim(object):
    def _get_optim(self, p):
        return self.pt_optim_constructor([p], **self._get_optim_args(p))

class PyroLRScheduler(PyroOptim):
    def _get_optim(self, p):
        optim = super(PyroLRScheduler, self)._get_optim(p)
        return self.pt_scheduler_constructor(optim, **self.kwargs)
```
and then replace this code with
```py
self.optim_objs[p] = self._get_optim(p)
```